File: backend/src/ingest/ip.ts
// backend/src/ingest/ip.ts
type Headers = Record<string, string | undefined>;

/**
 * Strip the port from a CloudFront-Viewer-Address value.
 * CloudFront always formats this header as `<ip>:<port>`, even for IPv6
 * (which is sent unbracketed, e.g. `2001:db8::1:443`).  Cutting after the
 * last colon is therefore correct for both address families.
 * X-Forwarded-For entries carry no port and are NOT passed through this function.
 */
function stripPort(value: string): string {
  const i = value.lastIndexOf(':');
  return i === -1 ? value : value.slice(0, i);
}

export function extractClientIp(headers: Headers): string {
  const viewer = headers['cloudfront-viewer-address'];
  if (viewer) return stripPort(viewer.trim());

  const xff = headers['x-forwarded-for'];
  if (xff) return xff.split(',')[0].trim();

  return 'unknown';
}

File: backend/internal/ip/ip.go
// Package ip ports the ingest/ip.ts client-IP extraction logic to Go.
package ip

import (
	"strings"
)

// stripPort removes the port suffix from value by cutting after the last colon.
// This is correct for both IPv4 ("198.51.100.10:46532" → "198.51.100.10") and
// unbracketed IPv6 as sent by CloudFront ("2001:db8::1:46532" → "2001:db8::1").
// If no colon is present, value is returned unchanged.
func stripPort(value string) string {
	i := strings.LastIndex(value, ":")
	if i == -1 {
		return value
	}
	return value[:i]
}

// ExtractClientIP reads the client IP from the request headers map.
// Header lookup treats an empty string as absent (mirrors JS truthiness).
//
// Priority:
//  1. cloudfront-viewer-address — CloudFront injects "<ip>:<port>"; IPv6 is
//     unbracketed, so we strip the last colon-delimited segment to obtain the
//     address.
//  2. x-forwarded-for — the first comma-separated entry is the originating IP;
//     no port is present in this header.
//  3. Falls back to "unknown".
func ExtractClientIP(headers map[string]string) string {
	if v := headers["cloudfront-viewer-address"]; v != "" {
		return stripPort(strings.TrimSpace(v))
	}
	if v := headers["x-forwarded-for"]; v != "" {
		parts := strings.SplitN(v, ",", 2)
		return strings.TrimSpace(parts[0])
	}
	return "unknown"
}
